src/archive/src/main.rs

```diff
@@ -0,0 +1,438 @@
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_entries(index: Option<u64>, limit: Option<u16>) -> Entries {
+    let num_entries = limit_or_default(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
+            Some(idx) => idx as usize,
+        };
+
+        let next_idx = if start_idx + num_entries < length {
+            Some((start_idx + num_entries) as u64)
+        } else {
+            None
+        };
+
+        let mut entries = Vec::with_capacity(num_entries);
+        for idx in start_idx..start_idx + num_entries {
+            let entry = match log.get(idx) {
+                None => break,
+                Some(entry) => entry,
+            };
+            entries.push(Some(
+                candid::decode_one(&entry).expect("failed to decode log entry"),
+            ))
+        }
+        Entries { entries, next_idx }
+    })
+}
+
+#[query]
+fn get_anchor_entries(anchor: Anchor, cursor: Option<Cursor>, limit: Option<u16>) -> AnchorEntries {
+    let num_entries = limit_or_default(limit);
+
+    with_anchor_index_mut(|index| {
+        let iterator = match cursor {
+            None => index.range(anchor.to_le_bytes().to_vec(), None),
+            Some(Cursor::NextToken { next_token }) => index.range(next_token.into_vec(), None),
+            Some(Cursor::Timestamp { timestamp }) => index.range(
+                anchor.to_le_bytes().to_vec(),
+                Some(timestamp.to_le_bytes().to_vec()),
+            ),
+        };
+
+        with_log(|log| {
+            let mut entries: Vec<(AnchorIndexKey, Vec<u8>)> = iterator
+                .take(num_entries + 1) // take one too many to extract the cursor
+                .map(|(anchor_key, _)| {
+                    let entry = log
+                        .get(anchor_key.log_index as usize)
+                        .expect("bug: index to non-existing entry");
+                    (anchor_key, entry)
+                })
+                .collect();
+
+            let cursor = if entries.len() > num_entries {
+                entries.pop().map(|(key, _)| Cursor::NextToken {
+                    next_token: ByteBuf::from(key.to_bytes()),
+                })
+            } else {
+                None
+            };
+
+            let entries = entries
+                .iter()
+                .map(|(_, entry)| candid::decode_one(&entry).expect("failed to decode log entry"))
+                .collect();
+
+            AnchorEntries { entries, cursor }
+        })
+    })
+}
+
+fn limit_or_default(limit: Option<u16>) -> usize {
+    with_config(|config| {
+        limit
+            .map(|l| l.min(config.max_entries_per_call))
+            .unwrap_or(config.max_entries_per_call) as usize
+    })
+}
+
+#[init]
+#[post_upgrade]
```

Comment:
> The interface for init and post_upgrade must be the same.

Interesting! How come? Is that a canister interface limitation, or a silly candid spec limitation?

Reply:
I think it's a bit of both:
* Candid cannot express the differing interfaces because it uses only one signature for services: `service: () -> {}`
* Some component in the stack relies on there being only one component, as I get a candid type error if I try to make the interfaces different.

Reply:
Gotcha! I think it could be interesting to create discussion on the SDK repo